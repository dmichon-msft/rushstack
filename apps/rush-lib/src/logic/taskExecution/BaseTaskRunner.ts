// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import type { StdioSummarizer } from '@rushstack/terminal';
import type { CollatedWriter } from '@rushstack/stream-collator';

import type { TaskStatus } from './TaskStatus';
import type { CommandLineConfiguration } from '../../api/CommandLineConfiguration';

export interface ITaskRunnerContext {
  repoCommandLineConfiguration: CommandLineConfiguration;
  collatedWriter: CollatedWriter;
  stdioSummarizer: StdioSummarizer;
  quietMode: boolean;
  debugMode: boolean;
}

/**
 * The `Task` class is a node in the dependency graph of work that needs to be scheduled by the
 * `TaskExecutionManager`. Each `Task` has a `runner` member of type `BaseTaskRunner`, whose subclass
 * manages the actual operations for running a single task.
 */
export abstract class BaseTaskRunner {
  /**
   * Name of the task definition.
   */
  public abstract readonly name: string;

  /**
   * This flag determines if the task is allowed to be skipped if up to date.
   */
  public abstract isSkipAllowed: boolean;

  /**
   * Assigned by execute().  True if the script was an empty string.  Operationally an empty string is
   * like a shell command that succeeds instantly, but e.g. it would be odd to report time statistics for it.
   */
  public abstract hadEmptyScript: boolean;

  /**
   * If set to true, a warning result should not make Rush exit with a nonzero
   * exit code
   */
  public abstract warningsAreAllowed: boolean;

  /**
   * Indicates if the output of this task may be written to the cache
   */
  public abstract isCacheWriteAllowed: boolean;

  /**
   * Method to be executed for the task.
   */
  public abstract executeAsync(context: ITaskRunnerContext): Promise<TaskStatus>;
}
