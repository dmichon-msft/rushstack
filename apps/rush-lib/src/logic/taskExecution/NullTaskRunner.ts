// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TaskStatus } from './TaskStatus';
import { ITaskRunner, ITaskRunnerContext } from './ITaskRunner';

/**
 * Implementation of `ITaskRunner` for tasks with empty scripts.
 */
export class NullTaskRunner implements ITaskRunner {
  private readonly _result: TaskStatus;
  public readonly name: string;
  public readonly hadEmptyScript: boolean = true;
  // The task may never be skipped; it doesn't do anything anyway
  public isSkipAllowed: boolean = false;
  // The task is a no-op, so skip writing an empty cache entry
  public isCacheWriteAllowed: boolean = false;
  // Nothing will get logged, no point allowing warnings
  public readonly warningsAreAllowed: boolean = false;

  public constructor(name: string, result: TaskStatus) {
    this.name = name;
    this._result = result;
  }

  public async executeAsync(context: ITaskRunnerContext): Promise<TaskStatus> {
    return this._result;
  }
}
