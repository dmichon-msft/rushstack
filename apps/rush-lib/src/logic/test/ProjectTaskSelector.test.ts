// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { RushConfiguration } from '../../api/RushConfiguration';
import { CommandLineConfiguration, IPhasedCommand } from '../../api/CommandLineConfiguration';
import { IProjectTaskOptions, IProjectTaskFactory, ProjectTaskSelector } from '../ProjectTaskSelector';
import { Task } from '../taskExecution/Task';
import { JsonFile } from '@rushstack/node-core-library';
import { ICommandLineJson } from '../../api/CommandLineJson';
import { RushConstants } from '../RushConstants';
import { TaskStatus } from '../taskExecution/TaskStatus';
import { MockTaskRunner } from '../taskExecution/test/MockTaskRunner';

interface ISerializedTask {
  name: string;
  isCacheWriteAllowed: boolean;
  dependencies: string[];
}

function serializeTask(task: Task): ISerializedTask {
  return {
    name: task.name,
    isCacheWriteAllowed: task.runner.isCacheWriteAllowed,
    dependencies: Array.from(task.dependencies, (dep: Task) => dep.name)
  };
}

describe('ProjectTaskSelector', () => {
  const rushJsonFile: string = `${__dirname}/workspaceRepo/rush.json`;
  const commandLineJsonFile: string = `${__dirname}/workspaceRepo/common/config/rush/command-line.json`;

  const taskFactory: IProjectTaskFactory = {
    createTask({ phase, project }: IProjectTaskOptions): Task {
      const name: string = `${project.packageName} (${phase.name.slice(
        RushConstants.phaseNamePrefix.length
      )})`;

      return new Task(new MockTaskRunner(name), TaskStatus.Ready);
    }
  };

  let rushConfiguration!: RushConfiguration;
  let commandLineConfiguration!: CommandLineConfiguration;

  beforeAll(() => {
    rushConfiguration = RushConfiguration.loadFromConfigurationFile(rushJsonFile);
    const commandLineJson: ICommandLineJson = JsonFile.load(commandLineJsonFile);

    commandLineConfiguration = new CommandLineConfiguration(commandLineJson);
  });

  describe('#createTasks', () => {
    it('handles a full build', () => {
      const buildCommand: IPhasedCommand = commandLineConfiguration.commands.get('build')! as IPhasedCommand;

      const taskSelector: ProjectTaskSelector = new ProjectTaskSelector({
        phases: commandLineConfiguration.phases,
        projects: rushConfiguration.projectsByName,
        phasesToRun: buildCommand.phases
      });

      // All projects
      expect(
        Array.from(
          taskSelector.createTasks({
            projectSelection: new Set(rushConfiguration.projects),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();
    });

    it('handles filtered projects', () => {
      const buildCommand: IPhasedCommand = commandLineConfiguration.commands.get('build')! as IPhasedCommand;

      const taskSelector: ProjectTaskSelector = new ProjectTaskSelector({
        phases: commandLineConfiguration.phases,
        projects: rushConfiguration.projectsByName,
        phasesToRun: buildCommand.phases
      });

      // Single project
      expect(
        Array.from(
          taskSelector.createTasks({
            projectSelection: new Set([rushConfiguration.getProjectByName('g')!]),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();

      // Filtered projects
      expect(
        Array.from(
          taskSelector.createTasks({
            projectSelection: new Set([
              rushConfiguration.getProjectByName('f')!,
              rushConfiguration.getProjectByName('a')!,
              rushConfiguration.getProjectByName('c')!
            ]),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();
    });

    it('handles filtered phases', () => {
      // Single phase with a missing dependency
      expect(
        Array.from(
          new ProjectTaskSelector({
            phases: commandLineConfiguration.phases,
            projects: rushConfiguration.projectsByName,
            phasesToRun: new Set([commandLineConfiguration.phases.get('_phase:upstream-self')!])
          }).createTasks({
            projectSelection: new Set(rushConfiguration.projects),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();

      // Two phases with a missing link
      expect(
        Array.from(
          new ProjectTaskSelector({
            phases: commandLineConfiguration.phases,
            projects: rushConfiguration.projectsByName,
            phasesToRun: new Set([
              commandLineConfiguration.phases.get('_phase:complex')!,
              commandLineConfiguration.phases.get('_phase:upstream-3')!,
              commandLineConfiguration.phases.get('_phase:upstream-1')!,
              commandLineConfiguration.phases.get('_phase:no-deps')!
            ])
          }).createTasks({
            projectSelection: new Set(rushConfiguration.projects),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();
    });

    it('handles filtered phases on filtered projects', () => {
      // Single phase with a missing dependency
      expect(
        Array.from(
          new ProjectTaskSelector({
            phases: commandLineConfiguration.phases,
            projects: rushConfiguration.projectsByName,
            phasesToRun: new Set([commandLineConfiguration.phases.get('_phase:upstream-2')!])
          }).createTasks({
            projectSelection: new Set([
              rushConfiguration.getProjectByName('f')!,
              rushConfiguration.getProjectByName('a')!,
              rushConfiguration.getProjectByName('c')!
            ]),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();

      // Phases with missing links
      expect(
        Array.from(
          new ProjectTaskSelector({
            phases: commandLineConfiguration.phases,
            projects: rushConfiguration.projectsByName,
            phasesToRun: new Set([
              commandLineConfiguration.phases.get('_phase:complex')!,
              commandLineConfiguration.phases.get('_phase:upstream-3')!,
              commandLineConfiguration.phases.get('_phase:upstream-1')!,
              commandLineConfiguration.phases.get('_phase:no-deps')!
            ])
          }).createTasks({
            projectSelection: new Set([
              rushConfiguration.getProjectByName('f')!,
              rushConfiguration.getProjectByName('a')!,
              rushConfiguration.getProjectByName('c')!
            ]),
            taskFactory
          }),
          serializeTask
        )
      ).toMatchSnapshot();
    });
  });
});
