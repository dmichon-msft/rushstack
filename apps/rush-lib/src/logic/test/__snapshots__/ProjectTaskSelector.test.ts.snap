// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTaskSelector #createTasks handles a full build 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "b (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "c (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "d (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "e (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "f (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "g (no-deps)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "b (no-deps)",
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "c (no-deps)",
      "b (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "d (no-deps)",
      "b (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "e (no-deps)",
      "c (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "f (no-deps)",
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "g (no-deps)",
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-self)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "b (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "b (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "c (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-2)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "d (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "e (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "d (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "e (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-2-self)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-3)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (complex)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-3)",
      "a (upstream-1-self-upstream)",
      "a (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (complex)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-3)",
      "b (upstream-1-self-upstream)",
      "b (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (complex)",
  },
  Object {
    "dependencies": Array [
      "d (upstream-3)",
      "b (upstream-1-self-upstream)",
      "b (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (complex)",
  },
  Object {
    "dependencies": Array [
      "e (upstream-3)",
      "c (upstream-1-self-upstream)",
      "c (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (complex)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-3)",
      "a (upstream-1-self-upstream)",
      "a (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (complex)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-3)",
      "a (upstream-1-self-upstream)",
      "a (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (complex)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered phases 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "a (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "b (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "d (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "e (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "f (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-self)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered phases 2`] = `
Array [
  Object {
    "dependencies": Array [
      "a (upstream-3)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (complex)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-3)",
    ],
    "isCacheWriteAllowed": false,
    "name": "b (complex)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-3)",
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (complex)",
  },
  Object {
    "dependencies": Array [
      "d (upstream-3)",
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "d (complex)",
  },
  Object {
    "dependencies": Array [
      "e (upstream-3)",
      "b (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "e (complex)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-3)",
    ],
    "isCacheWriteAllowed": false,
    "name": "f (complex)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-3)",
    ],
    "isCacheWriteAllowed": false,
    "name": "g (complex)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "b (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "d (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "b (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "e (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "f (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "b (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "b (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "b (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "d (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "c (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "e (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "b (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "c (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "d (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "e (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "f (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "g (no-deps)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered phases on filtered projects 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "f (upstream-2)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-2)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-2)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered phases on filtered projects 2`] = `
Array [
  Object {
    "dependencies": Array [
      "f (upstream-3)",
    ],
    "isCacheWriteAllowed": false,
    "name": "f (complex)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-3)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (complex)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-3)",
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (complex)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "f (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "f (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "c (no-deps)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered projects 1`] = `
Array [
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "g (no-deps)",
  },
  Object {
    "dependencies": Array [
      "g (no-deps)",
    ],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-self)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-2)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "g (upstream-2-self)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "g (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "g (upstream-3)",
    ],
    "isCacheWriteAllowed": false,
    "name": "g (complex)",
  },
]
`;

exports[`ProjectTaskSelector #createTasks handles filtered projects 2`] = `
Array [
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "f (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (no-deps)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "c (no-deps)",
  },
  Object {
    "dependencies": Array [
      "f (no-deps)",
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "c (no-deps)",
      "a (upstream-self)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-self)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-1)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-2)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-2)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-3)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-3)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-1)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-1-self)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-2)",
    ],
    "isCacheWriteAllowed": true,
    "name": "c (upstream-2-self)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-1-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [],
    "isCacheWriteAllowed": true,
    "name": "a (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "a (no-deps)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (upstream-1-self-upstream)",
  },
  Object {
    "dependencies": Array [
      "f (upstream-3)",
      "a (upstream-1-self-upstream)",
      "a (upstream-2-self)",
    ],
    "isCacheWriteAllowed": true,
    "name": "f (complex)",
  },
  Object {
    "dependencies": Array [
      "a (upstream-3)",
    ],
    "isCacheWriteAllowed": true,
    "name": "a (complex)",
  },
  Object {
    "dependencies": Array [
      "c (upstream-3)",
      "a (upstream-1-self)",
      "a (upstream-1)",
    ],
    "isCacheWriteAllowed": false,
    "name": "c (complex)",
  },
]
`;
